File: Client/Linux/lib/Platform/PersistentConnectionInternalClient.cpp
#include "PersistentConnectionInternalClient.h"
#include "../WebSocketConnection.h"
#include <cstdlib>

namespace Thriot { namespace Platform {

PersistentConnectionInternalClient::PersistentConnectionInternalClient(const string& url, WebSocketConnection* webSocketConnection)
{
	_url = url;
	_webSocketConnection = webSocketConnection;
	_isLoggedIn = false;
	_isSubscribed = false;
}

PlatformOperationResult PersistentConnectionInternalClient::Login(const string& deviceId, const string& apiKey)
{
	if(_isLoggedIn)
		return InvalidOperation;

	if(!_webSocketConnection->IsDisconnected())
		return ConnectedAlready;

	if(!_webSocketConnection->Connect(_url))
		return Disconnected;

	_webSocketConnection->OnMessage(this, OnMessagePayloadReceived);

	if(_webSocketConnection->IsDisconnected())
		return ConnectionFailed;

	_webSocketConnection->Send("login " + deviceId + " " + apiKey);

	string response = _webSocketConnection->WaitResponse(10);

	if(!response.empty())
	{
		if(response == "login ack")
		{
			_isLoggedIn = true;
			return Ok;
		}
		else
		{
			return LoginInvalid;
		}
	}
	else
	{
		return Timeout;
	}
}

PlatformOperationResult PersistentConnectionInternalClient::Subscribe(
				const SubscriptionType subscriptionType, OnMessageReceived onMessageReceived)
{
	if(!_isLoggedIn)
		return LoginRequired;

	if(_isSubscribed)
		return InvalidOperation;

	if(_webSocketConnection->IsDisconnected())
		return Disconnected;

	_subscriptionType = subscriptionType;
	if(subscriptionType == ReceiveAndForget)
	{
		_webSocketConnection->Send("subscribe receiveandforget");
	}
	else if(subscriptionType == PeekAndCommit)
	{
		_webSocketConnection->Send("subscribe peekandcommit");
	}
	else
	{
		return ArgumentOutOfRange;
	}

	string response = _webSocketConnection->WaitResponse(10);

	if(!response.empty())
	{
		if(response == "subscribe ack")
		{
			_isSubscribed = true;
			_onMessageReceived = onMessageReceived;
			return Ok;
		}

		if(response == "subscribe unauthorized")
			return LoginRequired;

		if(response == "subscribe already")
			return SubscribedAlready;

		return SubscribeInvalid;
	}
	else
	{
		return Timeout;
	}
}

PlatformOperationResult PersistentConnectionInternalClient::Unsubscribe()
{
	if(!_isLoggedIn)
		return LoginRequired;

	if(!_isSubscribed)
		return SubscriptionRequired;

	if(_webSocketConnection->IsDisconnected())
		return Disconnected;

	_webSocketConnection->Send("unsubscribe");
	
	string response = _webSocketConnection->WaitResponse(10);

	if(!response.empty())
	{
		if(response == "unsubscribe ack")
		{
			_isSubscribed = false;
			return Ok;
		}

		if(response == "unsubscribe unauthorized")
			return LoginRequired;

		if(response == "unsubscribe notsubscribed")
			return UnsubscribeNotsubscribed;

		return UnsubscribeInvalid;
	}
	else
	{
		return Timeout;
	}
}

PlatformOperationResult PersistentConnectionInternalClient::Close()
{
	if(!_isLoggedIn)
		return LoginRequired;

	_webSocketConnection->Send("close");

	_isLoggedIn = false;
	_isSubscribed = false;

	_webSocketConnection->Close();

	return Ok;
}

PlatformOperationResult PersistentConnectionInternalClient::RecordTelemetryData(const string& payload)
{
	if(!_isLoggedIn)
		return LoginRequired;

	if(_webSocketConnection->IsDisconnected())
		return Disconnected;

	_webSocketConnection->Send("telemetrydata " + payload);

	string response = _webSocketConnection->WaitResponse(10);

	if(!response.empty())
	{
		if(response == "telemetrydata ack")
			return Ok;

		if(response == "telemetrydata unauthorized")
			return LoginRequired;

		return TelemetryDataInvalid;
	}
	else
	{
		return Timeout;
	}
}

PlatformOperationResult PersistentConnectionInternalClient::SendMessageTo(const string& deviceId, const string& payload)
{
	if(!_isLoggedIn)
		return LoginRequired;

	if(_webSocketConnection->IsDisconnected())
		return Disconnected;

	_webSocketConnection->Send("sendto " + deviceId + " " + payload);

	string response = _webSocketConnection->WaitResponse(10);

	if(!response.empty())
	{
		if(response == "sendto ack")
			return Ok;

		if(response == "sendto unauthorized")
			return LoginRequired;

		return SendToInvalid;
	}
	else
	{
		return Timeout;
	}
}

PlatformOperationResult PersistentConnectionInternalClient::Heartbeat()
{
	if(!_isLoggedIn)
		return LoginRequired;

	if(_webSocketConnection->IsDisconnected())
		return Disconnected;

	_webSocketConnection->Send("heartbeat");

	string response = _webSocketConnection->WaitResponse(10);

	if(!response.empty())
	{
		if(response == "yo")
			return Ok;

		return HeartbeatInvalid;
	}
	else
	{
		return Timeout;
	}
}

void PersistentConnectionInternalClient::Spin()
{
	_webSocketConnection->Spin();
}

void PersistentConnectionInternalClient::HandlePushedMessage(const string& message)
{
	ProcessMessage(message);
	CommitIfNeeded();
}

void PersistentConnectionInternalClient::ProcessMessage(const string& message)
{
	// pushedmessage <messageid> <timestamp> <senderdeviceid> <payload>
	size_t afterCommandSeparator = message.find(' ');

	size_t afterMessageIdSeparator = message.find(' ', afterCommandSeparator + 1);
	if(afterMessageIdSeparator == string::npos)
		return;
	
	size_t afterTimestampSeparator = message.find(' ', afterMessageIdSeparator + 1);
	if(afterTimestampSeparator == string::npos)
		return;

	size_t afterSenderDeviceIdSeparator = message.find(' ', afterTimestampSeparator + 1);
	if(afterSenderDeviceIdSeparator == string::npos)
		return;

	string messageIdStr = message.substr(afterCommandSeparator + 1, afterMessageIdSeparator - afterCommandSeparator - 1);
	string timestampStr = message.substr(afterMessageIdSeparator + 1, afterTimestampSeparator - afterMessageIdSeparator - 1);
	string senderDeviceIdStr = message.substr(afterTimestampSeparator + 1, afterSenderDeviceIdSeparator - afterTimestampSeparator - 1);
	string payload = message.substr(afterSenderDeviceIdSeparator + 1);

	PushedMessage pushedMessage;
	pushedMessage.MessageId = (int)atol(messageIdStr.c_str());
	pushedMessage.Timestamp = atol(timestampStr.c_str());
	pushedMessage.SenderDeviceId = senderDeviceIdStr;
	pushedMessage.Payload = payload;

	_onMessageReceived(pushedMessage);
}

void PersistentConnectionInternalClient::CommitIfNeeded()
{
	if(_subscriptionType == PeekAndCommit)
	{
		_webSocketConnection->Send("commit");
	}
}

bool PersistentConnectionInternalClient::OnMessagePayloadReceived(const void* object, const string& message)
{
	PersistentConnectionInternalClient* pcic = (PersistentConnectionInternalClient*)object;

	if(message.find("pushedmessage ") == 0)
	{
		pcic->HandlePushedMessage(message);
		return true;
	}

	return false;
}
}}

File: Client/DotNet/Thriot.Client.DotNet/Platform/PersistentConnectionInternalClient.cs
using System;
using System.Threading;
using Thriot.Client.DotNet.Platform.Exceptions;

namespace Thriot.Client.DotNet.Platform
{
    internal class PersistentConnectionInternalClient
    {
        private readonly IWebSocketConnection _webSocketConnection;
        private bool _isLoggedIn;
        private bool _isSubscribed;
        private SubscriptionType _subscriptionType;
        
        private Action<PushedMessage> _onMessageReceived;
        
        private readonly AutoResetEvent _responseEvent;
        private string _lastResponse;

        public PersistentConnectionInternalClient(IWebSocketConnection webSocketConnection)
        {
            _webSocketConnection = webSocketConnection;
            _isLoggedIn = false;
            _isSubscribed = false;
            _responseEvent = new AutoResetEvent(false);
        }

        public void Login(string url, string deviceId, string apiKey)
        {
            if(_isLoggedIn)
                throw new InvalidOperationException("Already logged in");

            _webSocketConnection.Connect(url);
            _webSocketConnection.OnMessage(MessagePayloadReceived);

            if(_webSocketConnection.IsDisconnected)
                throw new DisconnectedException();

            _webSocketConnection.Send(string.Format("login {0} {1}", deviceId, apiKey));

            if (_responseEvent.WaitOne(TimeSpan.FromSeconds(10)))
            {
                if (_lastResponse == "login ack")
                {
                    _isLoggedIn = true;
                }
                else
                {
                    throw new LoginInvalidException(_lastResponse);
                }
            }
            else
            {
                throw new TimeoutException("Login timeout expired");
            }
        }

        public void Subscribe(SubscriptionType subscriptionType, Action<PushedMessage> onMessageReceived)
        {
            LoginRequired();

            if(_isSubscribed)
                throw new InvalidOperationException("Already subscribed");

            if (onMessageReceived == null)
                throw new ArgumentNullException("onMessageReceived");

            if (_webSocketConnection.IsDisconnected)
                throw new DisconnectedException();

            _subscriptionType = subscriptionType;
            if (subscriptionType == SubscriptionType.ReceiveAndForget)
            {
                _webSocketConnection.Send("subscribe receiveandforget");
            }
            else if (subscriptionType == SubscriptionType.PeekAndCommit)
            {
                _webSocketConnection.Send("subscribe peekandcommit");
            }
            else
            {
                throw new ArgumentOutOfRangeException("subscriptionType");
            }

            if (_responseEvent.WaitOne(TimeSpan.FromSeconds(10)))
            {
                if (_lastResponse == "subscribe ack")
                {
                    _isSubscribed = true;
                    _onMessageReceived = onMessageReceived;
                    return;
                }

                if (_lastResponse == "subscribe unauthorized")
                    throw new LoginRequiredException();

                if (_lastResponse == "subscribe already")
                    throw new SubscribedAlreadyException();

                throw new SubscribeInvalidException(_lastResponse);
            }
            else
            {
                throw new TimeoutException("Subscribe timeout expired");
            }
        }

        public void Unsubscribe()
        {
            LoginAndSubscriptionRequired();

            if (_webSocketConnection.IsDisconnected)
                throw new DisconnectedException();
            
            _webSocketConnection.Send("unsubscribe");

            if (_responseEvent.WaitOne(TimeSpan.FromSeconds(10)))
            {
                if (_lastResponse == "unsubscribe ack")
                {
                    _isSubscribed = false;
                    return;
                }

                if (_lastResponse == "unsubscribe unauthorized")
                    throw new LoginRequiredException();

                if (_lastResponse == "unsubscribe notsubscribed")
                    throw new UnsubscribeNotsubscribedException();

                throw new UnsubscribeInvalidException(_lastResponse);
            }
            else
            {
                throw new TimeoutException("Unsubscribe timeout expired");
            }
        }

        public void Close()
        {
            LoginRequired();
            
            _webSocketConnection.Send("close", false);

            _isLoggedIn = false;
            _isSubscribed = false;

            _webSocketConnection.Close();
        }

        public void RecordTelemetryData(string payload)
        {
            LoginRequired();

            if (_webSocketConnection.IsDisconnected)
                throw new DisconnectedException();
            
            _webSocketConnection.Send("telemetrydata " + payload);

            if (_responseEvent.WaitOne(TimeSpan.FromSeconds(10)))
            {
                if (_lastResponse == "telemetrydata ack")
                    return;

                if (_lastResponse == "telemetrydata unauthorized")
                    throw new LoginRequiredException();

                throw new TelemetryDataInvalidException(_lastResponse);
            }
            else
            {
                throw new TimeoutException("RecordTelemetryData timeout expired");
            }
        }

        public void SendMessageTo(string deviceId, string payload)
        {
            LoginRequired();

            if (_webSocketConnection.IsDisconnected)
                throw new DisconnectedException();
            
            _webSocketConnection.Send("sendto " + deviceId + " " + payload);

            if (_responseEvent.WaitOne(TimeSpan.FromSeconds(10)))
            {
                if (_lastResponse == "sendto ack")
                    return;

                if (_lastResponse == "sendto unauthorized")
                    throw new LoginRequiredException();

                throw new SendToInvalidException(_lastResponse);
            }
            else
            {
                throw new TimeoutException("Sendto timeout expired");
            }
        }

        private void MessagePayloadReceived(string msg, WebSocketConnection connection)
        {
            if (msg == "yo")
            {
                try
                {
                    _webSocketConnection.Send("heartbeat");
                }
                catch
                {
                    // if not succeeds then may reconnect
                }
                return;
            }
            if (msg.StartsWith("pushedmessage "))
            {
                HandlePushedMessage(msg);
                return;
            }

            _lastResponse = msg;
            _responseEvent.Set();
        }

        private void HandlePushedMessage(string msg)
        {
            ProcessMessage(msg);

            CommitIfNeded();
        }

        private void ProcessMessage(string msg)
        {
            var parts = msg.Split(new[] {' '}, 5);

            if (parts.Length < 3)
                throw new ArgumentException("pushedMessage");

            var messageIdText = parts[1];
            int messageId;

            if (!int.TryParse(messageIdText, out messageId))
                throw new ArgumentException("Invalid messageid");

            var timestampText = parts[2];
            long timestamp;

            if (!long.TryParse(timestampText, out timestamp))
                throw new ArgumentException("Invalid timespan");

            string senderDeviceId = parts[3];

            var messageText = msg.Substring(parts[0].Length + parts[1].Length + parts[2].Length + parts[3].Length + 4);

            _onMessageReceived(new PushedMessage(messageId, timestamp, messageText, senderDeviceId));
        }

        private void CommitIfNeded()
        {
            if (_subscriptionType == SubscriptionType.PeekAndCommit)
            {
                try
                {
                    _webSocketConnection.Send("commit");
                }
                catch
                {
                    // if not succeeds then peek the message again, do not care
                }
            }
        }

        private void LoginAndSubscriptionRequired()
        {
            if (!_isLoggedIn)
                throw new LoginRequiredException();

            if (!_isSubscribed)
                throw new SubscriptionRequiredException();
        }

        private void LoginRequired()
        {
            if (!_isLoggedIn)
                throw new LoginRequiredException();
        }
    }
}
